compiler/rustc_hir_analysis/src/structured_errors/wrong_number_of_generic_args.rs

```diff
@@ -758,7 +758,7 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
 
         let trait_generics = self.tcx.generics_of(trait_);
         let num_trait_generics_except_self =
-            trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
+            trait_generics.count() - trait_generics.has_self as usize;
```

Comment:
I'm not sure whether this really is better.. I read the bottom one as an `if` still, which is now just less obvious. We also suddenly have to care about `usize`, while in the top one the integer type is implicit and I don't have to care about it at all.

Basically, this change makes the `if` implicit and the integer type explicit, exactly the wrong way around :(

Reply:
The reason I like the cast is that it reads as a "just type conversion", the `bool` already has the right value and I feel like the `if` obfuscates this. But I also see your point so :shrug:

Reply:
So I see this conversion as more of a "cool bit hack", which should not be in high level code worrying about skipping the right number of arguments. Especially since here has_self does only coincidentally represent the number of self types to skip, its primary purpose is representing whether there's a self type.

Reply:
Some platforms choose to implement true as all 1, I think there's even some platform that uses true=0 false=1 (which is where UNIX style exit codes come from!) This further shows that Rust using true=1 false=0 is kinda arbitrary and mostly a matter of representation, not something intrinsic to bools.

Reply:
"bool" is already deeply subjective itself: I regularly interact with code that uses `is_this_value_absent: bool` which means using `true` to indicate **non-presence**.

Reply:
Interesting, thanks!

I was thinking of `bool` as a `u1`, with maybe added semantics of `false` and `true`. But clearly this is just me.

Reply:
don't worry waffle i will still love you
![pleading heart eyes](https://cdn.discordapp.com/emojis/1043490107619020831.webp?size=96&quality=lossless)

Reply:
///////